File: packages/frontdmi_10a/lib/models/paquete_model.dart
import 'package:cloud_firestore/cloud_firestore.dart';

// Modelo de Paquete - BojitaNoir
class Paquete {
  final String id;
  final String destinatario;
  // final String direccion; // ELIMINADO: Reemplazado por origen y destino
  final Map<String, dynamic> origen;
  final Map<String, dynamic> destino;
  final double peso;
  final String estado; // pendiente, asignado, en_transito, entregado.
  final String? fotoUrl;
  final DateTime fechaCreacion;
  final String? repartidorId;
  final String? clienteId;
  final String? codigoQR;
  final Map<String, dynamic>? ubicacionRecoleccion; // {lat, lng, timestamp}
  final Map<String, dynamic>? ubicacionEntrega; // {lat, lng, timestamp}

  Paquete({
    required this.id,
    required this.destinatario,
    required this.origen,
    required this.destino,
    required this.peso,
    required this.estado,
    this.fotoUrl,
    required this.fechaCreacion,
    this.repartidorId,
    this.clienteId,
    this.codigoQR,
    this.ubicacionRecoleccion,
    this.ubicacionEntrega,
  });

  // Serialización desde JSON
  factory Paquete.fromJson(Map<String, dynamic> json) {
    return Paquete(
      id: json['id'] ?? '',
      destinatario: json['destinatario'] ?? '',
      origen: json['origen'] != null
          ? Map<String, dynamic>.from(json['origen'])
          : {
              'direccion': json['direccion'] ?? '',
            }, // Fallback para datos viejos
      destino: json['destino'] != null
          ? Map<String, dynamic>.from(json['destino'])
          : {
              'direccion': json['direccion'] ?? '',
            }, // Fallback para datos viejos
      peso: (json['peso'] ?? 0).toDouble(),
      estado: json['estado'] ?? 'pendiente',
      fotoUrl: json['fotoUrl'],
      fechaCreacion: json['fechaCreacion'] != null
          ? (json['fechaCreacion'] as Timestamp).toDate()
          : DateTime.now(),
      repartidorId: json['repartidorId'],
      clienteId: json['clienteId'],
      codigoQR: json['codigoQR'],
      ubicacionRecoleccion: json['ubicacionRecoleccion'] != null
          ? Map<String, dynamic>.from(json['ubicacionRecoleccion'])
          : null,
      ubicacionEntrega: json['ubicacionEntrega'] != null
          ? Map<String, dynamic>.from(json['ubicacionEntrega'])
          : null,
    );
  }

  // Serialización a JSON
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'destinatario': destinatario,
      'origen': origen,
      'destino': destino,
      'peso': peso,
      'estado': estado,
      'fotoUrl': fotoUrl,
      'fechaCreacion': Timestamp.fromDate(fechaCreacion),
      'repartidorId': repartidorId,
      'clienteId': clienteId,
      'codigoQR': codigoQR,
      'ubicacionRecoleccion': ubicacionRecoleccion,
      'ubicacionEntrega': ubicacionEntrega,
    };
  }

  // Copia con modificaciones
  Paquete copyWith({
    String? id,
    String? destinatario,
    Map<String, dynamic>? origen,
    Map<String, dynamic>? destino,
    double? peso,
    String? estado,
    String? fotoUrl,
    DateTime? fechaCreacion,
    String? repartidorId,
    String? clienteId,
    String? codigoQR,
    Map<String, dynamic>? ubicacionRecoleccion,
    Map<String, dynamic>? ubicacionEntrega,
  }) {
    return Paquete(
      id: id ?? this.id,
      destinatario: destinatario ?? this.destinatario,
      origen: origen ?? this.origen,
      destino: destino ?? this.destino,
      peso: peso ?? this.peso,
      estado: estado ?? this.estado,
      fotoUrl: fotoUrl ?? this.fotoUrl,
      fechaCreacion: fechaCreacion ?? this.fechaCreacion,
      repartidorId: repartidorId ?? this.repartidorId,
      clienteId: clienteId ?? this.clienteId,
      codigoQR: codigoQR ?? this.codigoQR,
      ubicacionRecoleccion: ubicacionRecoleccion ?? this.ubicacionRecoleccion,
      ubicacionEntrega: ubicacionEntrega ?? this.ubicacionEntrega,
    );
  }
}

File: packages/paqueteria/src/main/java/mx/edu/utez/paqueteria/model/PaqueteModel.java
package mx.edu.utez.paqueteria.model;

import com.google.cloud.Timestamp;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.util.Map;

@Data
@NoArgsConstructor
@AllArgsConstructor
public class PaqueteModel {
    private String id;
    private String destinatario;

    // Direcciones detalladas (origen y destino)
    // Campos: calle, numero, cp, colonia, municipio, estado, referencias, telefono,
    // lat, lng
    private Map<String, Object> origen;
    private Map<String, Object> destino;

    private Double peso;
    private String estado; // pendiente, asignado, en_transito, entregado
    private String clienteId;
    private String repartidorId;
    private String fotoUrl;
    private Timestamp fechaCreacion;
    private String codigoQR;

    // Ubicaciones de escaneo
    // Campos: lat, lng, timestamp
    private Map<String, Object> ubicacionRecoleccion;
    private Map<String, Object> ubicacionEntrega;

    // Getters y Setters
    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getDestinatario() {
        return destinatario;
    }

    public void setDestinatario(String destinatario) {
        this.destinatario = destinatario;
    }

    public Map<String, Object> getOrigen() {
        return origen;
    }

    public void setOrigen(Map<String, Object> origen) {
        this.origen = origen;
    }

    public Map<String, Object> getDestino() {
        return destino;
    }

    public void setDestino(Map<String, Object> destino) {
        this.destino = destino;
    }

    public Double getPeso() {
        return peso;
    }

    public void setPeso(Double peso) {
        this.peso = peso;
    }

    public String getEstado() {
        return estado;
    }

    public void setEstado(String estado) {
        this.estado = estado;
    }

    public String getClienteId() {
        return clienteId;
    }

    public void setClienteId(String clienteId) {
        this.clienteId = clienteId;
    }

    public String getRepartidorId() {
        return repartidorId;
    }

    public void setRepartidorId(String repartidorId) {
        this.repartidorId = repartidorId;
    }

    public String getFotoUrl() {
        return fotoUrl;
    }

    public void setFotoUrl(String fotoUrl) {
        this.fotoUrl = fotoUrl;
    }

    public Timestamp getFechaCreacion() {
        return fechaCreacion;
    }

    public void setFechaCreacion(Timestamp fechaCreacion) {
        this.fechaCreacion = fechaCreacion;
    }

    public String getCodigoQR() {
        return codigoQR;
    }

    public void setCodigoQR(String codigoQR) {
        this.codigoQR = codigoQR;
    }

    public Map<String, Object> getUbicacionRecoleccion() {
        return ubicacionRecoleccion;
    }

    public void setUbicacionRecoleccion(Map<String, Object> ubicacionRecoleccion) {
        this.ubicacionRecoleccion = ubicacionRecoleccion;
    }

    public Map<String, Object> getUbicacionEntrega() {
        return ubicacionEntrega;
    }

    public void setUbicacionEntrega(Map<String, Object> ubicacionEntrega) {
        this.ubicacionEntrega = ubicacionEntrega;
    }

    /**
     * Método helper para obtener dirección legible desde el mapa destino
     * Útil para notificaciones y logs
     */
    public String getDireccionFormateada() {
        if (destino == null)
            return "";

        StringBuilder sb = new StringBuilder();
        if (destino.get("calle") != null)
            sb.append(destino.get("calle"));
        if (destino.get("numero") != null)
            sb.append(" ").append(destino.get("numero"));
        if (destino.get("colonia") != null)
            sb.append(", ").append(destino.get("colonia"));
        if (destino.get("municipio") != null)
            sb.append(", ").append(destino.get("municipio"));

        return sb.toString().trim();
    }
}
